test/EngineTest.cpp

```diff
@@ -23,114 +32,259 @@ auto I = [](const auto& id) {
   return Id::makeFromVocabIndex(VocabIndex::make(id));
 };
 
-TEST(EngineTest, joinTest) {
-  IdTable a(2, allocator());
-  a.push_back({I(1), I(1)});
-  a.push_back({I(1), I(3)});
-  a.push_back({I(2), I(1)});
-  a.push_back({I(2), I(2)});
-  a.push_back({I(4), I(1)});
-  IdTable b(2, allocator());
-  b.push_back({I(1), I(3)});
-  b.push_back({I(1), I(8)});
-  b.push_back({I(3), I(1)});
-  b.push_back({I(4), I(2)});
-  IdTable res(3, allocator());
-  int lwidth = a.numColumns();
-  int rwidth = b.numColumns();
-  int reswidth = a.numColumns() + b.numColumns() - 1;
-  Join J{Join::InvalidOnlyForTestingJoinTag{}};
-  CALL_FIXED_SIZE((std::array{lwidth, rwidth, reswidth}), &Join::join, J, a, 0,
-                  b, 0, &res);
-
-  ASSERT_EQ(I(1), res(0, 0));
-  ASSERT_EQ(I(1), res(0, 1));
-  ASSERT_EQ(I(3), res(0, 2));
-  ASSERT_EQ(I(1), res(1, 0));
-  ASSERT_EQ(I(1), res(1, 1));
-  ASSERT_EQ(I(8), res(1, 2));
-  ASSERT_EQ(I(1), res(2, 0));
-  ASSERT_EQ(I(3), res(2, 1));
-  ASSERT_EQ(I(3), res(2, 2));
-  ASSERT_EQ(I(1), res(3, 0));
-  ASSERT_EQ(I(3), res(3, 1));
-  ASSERT_EQ(I(8), res(3, 2));
-  ASSERT_EQ(5u, res.size());
-  ASSERT_EQ(I(4), res(4, 0));
-  ASSERT_EQ(I(1), res(4, 1));
-  ASSERT_EQ(I(2), res(4, 2));
-
-  res.clear();
-  for (size_t i = 1; i <= 10000; ++i) {
-    b.push_back({I(4 + i), I(2 + i)});
+/*
+ * Return an 'IdTable' with the given 'tableContent'. all rows must have the
+ * same length.
+*/
+IdTable makeIdTableFromVector(std::vector<std::vector<size_t>> tableContent) {
+  IdTable result(tableContent[0].size(), allocator());
+
+  // Copying the content into the table.
+  for (const auto& row: tableContent) {
+    const size_t backIndex = result.size();
+    result.emplace_back();
+
+    for (size_t c = 0; c < tableContent[0].size(); c++) {
+      result(backIndex, c) = I(row[c]);
+    }
   }
-  a.push_back({I(400000), I(200000)});
-  b.push_back({I(400000), I(200000)});
 
-  CALL_FIXED_SIZE((std::array{lwidth, rwidth, reswidth}), &Join::join, J, a, 0,
-                  b, 0, &res);
-  ASSERT_EQ(6u, res.size());
+  return result;
+}
 
-  a.clear();
-  b.clear();
-  res.clear();
+/*
+ * A structure containing all information needed for a normal join test. A
+ * normal join test is defined as two IdTables being joined, the resulting
+ * IdTable tested, if it is sorted after the join column, or not, and this
+ * IdTable than compared with the sampleSolution.
+ */
+struct normalJoinTest {
+  IdTable leftIdTable;
+  size_t leftJC; // Join column for the left IdTable.
+  IdTable rightIdTable;
+  size_t rightJC; // Join column for the right IdTable.
+  bool testForSorted; // Should the result of the join be tested, if it is
+                      // orderd after leftJC? In other words, ordered after the join column?
+  IdTable sampleSolution;
```

Comment:
> And I am wondering, Do we really need the "testForSorted", or don't we rather want the information `leftIsSortedByJoinColumn` and the same for right? Because from this information we can distill all the cases for the hash and zipper join.

I don't quite follow the logic of this comment. What use would we have from knowing, if the input was sorted? A test case defined by the construct here only describes a normal join case, it (currently) doesn't dictate, for which join algorithm the test case is for.  

I could add another helper function, that uses join functions based on if the entries are sorted, but I would rather let the `TEST(...)`s decide what functions to use, than hard code them in the helper functions. More flexibility like that and easier to adjust if, for example, a new join algorithm gets added.

Reply:
P.S.: I changed the variable names as recommended in [48beb5e](https://github.com/ad-freiburg/qlever/pull/821/commits/48beb5eb876ea34a2669e7dfddb54f31baf84d7a).